### ISSUE TITLE
Tagging / Notification Support

### ISSUE THREAD

**Issue by dandesousa** [PokeNavBot/Issue-Tracker#90]

**Comment by fortepc** [comment]
Idea:

Step 1) User 1: `$subscribe registeel`
Step 2) PokeNav adds Registeel to a list of pokemon that User1 wants notified of. 
Step 3) User 2: `$raid registeel` 
Step 4) PokeNav uses some method to notify User 1 of that raid being reported. 


Step 2 Options) 

A -  Keep the subscribed pokemon as a profile feature. Having the ability to have cross server notification support. This would be better if users are able to exclude specific servers from pokemon subscriptions
B - Keep the subscribed pokemon as a server role. Making it server specific(like meowth)

Step 4 Options) 
A - DM the user with a link to join that raid channel.
B - Tag the user, either directly or with a role

**Comment by cupkait** [comment]
To maximize customization options, allow servers to assign server-specific roles/subscriptions to specific gyms and raid bosses.

For example, the ability to assign a sponsored gym with a role such as "EX HUNTER" and every time a raid is called out for that gym, the assigned role is also included and notified.

On the flip side of that with raid bosses, allowing servers to assign @Legendary, @Regirock, @Marowak, @Shiny to individual bosses so that these roles will be notified when the raid is called out.

**Comment by dandesousa** [comment]
@fortepc @cupkait 

Here we go, the most flexible solution I can think up this morning:

**Proposal**

*Challenges*

Discord is not designed for the flexibility that many people currently use. It can't scale to notify you of all Pokemon, no matter how much people want it to. Allowing subscribing to Pokemon willy nilly will flood discord servers with roles that aren't ever notified. And often the way things should be tagged is not necessarily obvious or enumerable (ex gym, park everyone plays at, poke draft bosses, legendaries, special sub-group etc).

Role tagging will become more important when in app screenshot reports become common and the app may not have access to all roles for manual tagging by a user.

*Goal*

Allow PokeNav to tag specific roles when certain events occur, which should save manual effort. This aims to allow admins to configure tags when a boss raid is announced or updated; or a raid occurs at a certain gym. It will solve the following example cases:

- User posts a Mewtwo raid, PokeNav can tag a `mewtwo` role to notify users with permissions to view the raid channel.
- User posts a Claydol raid, PokeNav can notify a `poke-draft-raid-boss` and `tier-3` role
- Someone posts a raid at "Bradford Dam Trail", an ex eligible gym. PokeNav can notify an `ex-eligible-raid` role.

**Implementation & Examples**

*Admin Commands*
`$create-role name-of-role` (convenience so you don't need another bot to use the feature)
`$enable-gym-role ex-downtown-role "Clock Tower"`
`$enable-boss-role pokedraft-raid-boss Claydol`
`$enable-boss-role legendary Mewtwo`
`$enable-boss-role mewtwo Mewtwo`

*User Commands*
`$join-role ex-downtown-role` (convenience -- allows joining any role assigned below)
`$leave-role ex-downtown-role` (convenience -- allows leaving any role assigned below)
`$show-roles` (will show a settings like embed with the boss and gym roles listed -- available to all)

*Restrictions / Limits*

- There are discord role limits in each server (250), so this is a hard upper limit I can't change.
- Roles will never be made automatically by members, only by admins / mods (recipe for disaster).
- Only raids will trigger notifications.
- When a user posts a raid, it will notify up to N (lets say 5) roles in the announcement message in the lobby. If the gym or boss wasn't set, and becomes set, it will post **at most one** message in the lobby with the appropriate tag. If a follow up notification was already sent to the lobby, another won't be sent for that raid.

**Questions / Future Work**

I will likely take the first simplest implementation to start but possible improvements include:

- Restrictions on roles based on time of day (edge case but consider a group that only raids at lunch, between 11-1pm).
- Should we bother with follow up raid notifications (it seems useful but not sure)

**Comment by dandesousa** [comment]
First, I think the active raid bosses thing is out of scope since bosses are regional and thats not what I'm trying to spend my time on with this feature necessarily. Unless it uses the "has it ever been a raid boss" feature used for counter lookups, the auto-create role from active boss probably wouldn't be implemented now.

Second, While I know you really like Meowth's mode of auto-creating roles by users, I think that its  dangerous, as it essentially grants anyone with send messages the ability anyone pokenav exists to create as many roles as they want on a server, regardless of utility. While you can somewhat do this now with raids those are temporary and immediately obvious to everyone whats going on. Creation of roles has a tendency to silently sneak up and when its a problem (run out of roles) requires immediate attention since it break half the features in PokeNav. Then the admin tags me in #support asking why PokeNav suddenly stopped working when a user typed a command.

In either case the above is purely additive, we can break it off into another feature and add it later as an option thats enabled.

As a simplifying step, I also don't see a need to treat boss roles as any different than gym roles in terms of command structure. If user don't subscribe to pokemon or gyms and instead subscribe to the roles on the server they want, admins gain **much** more flexibility in running servers how they want. 

`$subscribe legendary`
`$subscribe snorlax` 

Where the role is one that is pre-made (whitelisted) by admins.

Admins can change the roles and people don't need to subscribe, (make a legendary tag, and admins switch over the tag to the newest legendary t5 boss). PokeNav can still provide tools in the future to make reconfiguring the server easier (like if we allow the active raid bosses thing or swap to the next legendaries). The interface to users should remain the same. They lookup what notifications are enabled on the server, if they want they request new roles from admins and admins are the only ones that allow server configuration to change, since they are the only ones who can fix it.

**Comment by fortepc** [comment]
1) The point of spending your time keeping up with raid bosses is fair, I wish we could hook into other system like this page https://www.pokebattler.com/raids where they are already maintaining what current possible raid bosses are. PokeBattler seems to be pretty API friendly. Maybe since they already actively keep up with what active raid bosses are they have an API or would be willing to add one?

**Comment by fortepc** [comment]
2) And that makes sense too, but I am just thinking of the smaller communities. The ones that are less tech savy. And seeing that it _does_ work for Meowth servers. They've not had any need to delete that feature from their bot. It's fairly self regulating. I know that PokeNav does rely on roles a lot more than Meowth does, so it stands to fail a lot more drastically if An anarchy mode for subscriptions stands to benefit those smaller servers who are less tech savvy who wouldn't bother with the admin work needed to keep and maintain a set of roles for subscriptions.

**Comment by dandesousa** [comment]
I think a follow up issue to track the current bosses from pokebattler or another source would be good.

I do have one more change to this before I start the implementation if @fortepc / @cupkait can give me a thumbs up if they agree I'll proceed to work on it right away:

Theres one other case that this doesn't solve is when you wanna tag based on combination of things.

For example, If you want a role for tier 5's in a specific neighborhood or at a specific gym. Two separate problems. I propose an area rule in addition to gyms and the ability to combine a gym with a boss / tier. To future proof, align it with the trade syntax:

*EX Raid Role*
`$enable-raid-role ex-eligible-raid "gym: Ex Gym Name"`

*All tier 5 eggs and bosses*
`$enable-raid-role t5-raid "tier: 5"`

*All mewtwo*
`$enable-raid-role mewtwo-raid "boss: mewtwo"`

*All tier 5's within 5 KM of a GPS coordinate*
`$enable-raid-role township-raid "tier: 5" "area: 40.22, -75.01, 5"`

`tier` and `boss` would be mutually exclusive, as would `area` and `gym`.

There would probably also be limits on the number of rules that are allowed to be evaluated (I need to do some investigation).

To sum up, I see 4-5 new commands:

`subscribe <role-name>`. # subscribe to any role that is a raid role in the server
`enable-raid-role <role-name> <rules>`. # this gives you an id you can check
`disable-raid-role <an id>` # removes the rule
`show-subscriptions` # display everything in English language
`create-role role-name` # convenience to make a new role that is mentionable
